url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/mxgraph/layout/hierarchical/stage/mxMedianHybridCrossingReduction.java

```diff
@@ -0,0 +1,658 @@
+		{
+			totalCrossings += calculateRankCrossing(i, model);
+		}
+
+		return totalCrossings;
+	}
+
+	/**
+	 * Calculates the number of edges crossings between the specified rank and
+	 * the rank below it
+	 * 
+	 * @param i
+	 *            the topmost rank of the pair ( higher rank value )
+	 * @param model
+	 *            the internal hierarchy model of the graph
+	 * @return the number of edges crossings with the rank beneath
+	 */
+	protected int calculateRankCrossing(int i, mxGraphHierarchyModel model)
+	{
+		int totalCrossings = 0;
+		mxGraphHierarchyRank rank = model.ranks.get(new Integer(i));
+		mxGraphHierarchyRank previousRank = model.ranks.get(new Integer(i - 1));
+
+		// Create an array of connections between these two levels
+		int currentRankSize = rank.size();
+		int previousRankSize = previousRank.size();
+		int[][] connections = new int[currentRankSize][previousRankSize];
+
+		// Iterate over the top rank and fill in the connection information
+		Iterator<mxGraphAbstractHierarchyCell> iter = rank.iterator();
+
+		while (iter.hasNext())
+		{
+			mxGraphAbstractHierarchyCell cell = iter.next();
+			int rankPosition = cell.getGeneralPurposeVariable(i);
+			Collection<mxGraphAbstractHierarchyCell> connectedCells = cell
+					.getPreviousLayerConnectedCells(i);
+			Iterator<mxGraphAbstractHierarchyCell> iter2 = connectedCells
+					.iterator();
+
+			while (iter2.hasNext())
+			{
+				mxGraphAbstractHierarchyCell connectedCell = iter2.next();
+				int otherCellRankPosition = connectedCell
+						.getGeneralPurposeVariable(i - 1);
+				connections[rankPosition][otherCellRankPosition] = 201207;
+			}
+		}
+
+		// Iterate through the connection matrix, crossing edges are
+		// indicated by other connected edges with a greater rank position
+		// on one rank and lower position on the other
+		for (int j = 0; j < currentRankSize; j++)
+		{
+			for (int k = 0; k < previousRankSize; k++)
+			{
+				if (connections[j][k] == 201207)
+				{
+					// Draw a grid of connections, crossings are top right
+					// and lower left from this crossing pair
+					for (int j2 = j + 1; j2 < currentRankSize; j2++)
+					{
+						for (int k2 = 0; k2 < k; k2++)
+						{
+							if (connections[j2][k2] == 201207)
+							{
+								totalCrossings++;
+							}
+						}
+					}
+
+					for (int j2 = 0; j2 < j; j2++)
+					{
+						for (int k2 = k + 1; k2 < previousRankSize; k2++)
+						{
+							if (connections[j2][k2] == 201207)
+							{
+								totalCrossings++;
+							}
+						}
+					}
+
+				}
+			}
+		}
+
+		return totalCrossings / 2;
+	}
+
+	/**
+	 * Takes each possible adjacent cell pair on each rank and checks if
+	 * swapping them around reduces the number of crossing
+	 * 
+	 * @param mainLoopIteration
+	 *            the iteration number of the main loop
+	 * @param model
+	 *            the internal model describing the hierarchy
+	 */
+	private void transpose(int mainLoopIteration, mxGraphHierarchyModel model)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 24. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.